### PR TITLE
Allowing heterogeneous meanings for datastore list properties.

### LINKED ISSUE #1649: Different meanings set on values when reading null
Hey,
we have a list in our datastore with ids and sometimes nulls in it. Like here:
```
["null","null","a953c730fc2e4647b3a5b4e88fa4f0ce","59ce754ddf03401eb09abd50c977cf67","null"]	
```

When we try to fetch them, we get the "Different meanings set on values within a list_value" Exception.

Is this a bug, that the method does not handles null?

[`_get_meaning` Method][1]:

```python
def _get_meaning(value_pb, is_list=False):
    meaning = None
    if is_list:
        # An empty list will have no values, hence no shared meaning
        # set among them.
        if len(value_pb.list_value) == 0:
            return None

        # We check among all the meanings, some of which may be None,
        # the rest which may be enum/int values.
        all_meanings = set(_get_meaning(sub_value_pb)
                           for sub_value_pb in value_pb.list_value)
        meaning = all_meanings.pop()
        # The value we popped off should have been unique. If not
        # then we can't handle a list with values that have more
        # than one meaning.
        if all_meanings:
            raise ValueError('Different meanings set on values '
                             'within a list_value')
    elif value_pb.meaning:  # Simple field (int32)
        meaning = value_pb.meaning

    return meaning
```

[1]: https://github.com/GoogleCloudPlatform/gcloud-python/blob/efc1d41d230330fdd2f3c5c8df493b1ffb654be5/gcloud/datastore/helpers.py#L73

### PR DESCRIPTION
**Pull request by dhermes**
Fixes #1649.

@pcostell PTAL

### CODE REVIEW

**Comment by dhermes** [comment]
Running my [repro snippet][1] after this change verifies that proto -> Entity -> proto doesn't drop any infromation on the floor.

```python
# python pb_repro.py
e._meanings:
{u'blocks.value': ([None, 15, 15, None], [u'oops', u'foo', u'bar', u'forgot'])}
pb_new == pb:
True
```

[1]: https://gist.github.com/dhermes/054fac0d3a4147ac96d5

File: gcloud/datastore/helpers.py [hunk 1]
```diff
@@ -155,6 +155,48 @@ def entity_from_protobuf(pb):
...
+        return
+
+    meaning, orig_value = entity._meanings[name]
+    # Only add the meaning back to the protobuf if the value is
+    # unchanged from when it was originally read from the API.
+    if orig_value is not value:
+        return
+
+    # For lists, we set meaning on each sub-element.
+    if is_list:
```
Comment (pcostell) [inline gcloud/datastore/helpers.py]: I wonder if we want change detection on the individual elements too?

Comment (dhermes) [inline gcloud/datastore/helpers.py]: This isn't the detection step, this takes the values cached on the `Entity` and puts them back when creating a protobuf.

Also, how would detection differ on individual `Value` protobuf objects? How could it be different from `value.meaning`?

Comment (pcostell) [inline gcloud/datastore/helpers.py]: IIUC, if we see a value A, we'll only preserve the meaning if the value was set to A before. 

However, if we had [A, B] before, then we got [A, C], we'll drop the meaning for both A and C. However, couldn't we detect that A is still the same, and preserve the meaning? Basically, don't return above in the event of a list and instead try to preserve things?

Note: I think this is already much nicer than before, so I don't think this is necessary until users start to hit it. Just wanted to point it out as somewhere we could maybe make things a little nicer.

Comment (dhermes) [inline gcloud/datastore/helpers.py]: Gotcha. The reasoning behind why it was `[A, B]` or nothing was that meanings were completely foreign outside of GAE so modifying a property that had a meaning would drop all that information.

We could track individual elements, but it is almost certainly more trouble than it's worth. The check `orig_value is not value` means literally the same object. A `list` is mutable though, so users could make it shorter or longer or move items into a different order without making it a different object. Hence our decision to just "assume" that no changes means no changes.

**Comment by dhermes** [comment]
@tseaver PTAL.

@pcostell Is this good from your end?

**Comment by pcostell** [comment]
Yup, LGTM

**Comment by tseaver** [comment]
Given that we've decided to incorporate the notion of heterogenous meanings at all (a feature which seems quite dubious to me), the implementation looks good to merge.

**Comment by dhermes** [comment]
@tseaver Yes I feel like the `meaning`s code is a very steep uphill climb.

**Pull request merged**